### PR TITLE
Transformers custom handler example (CoLA Dataset)

### PR DESCRIPTION
**Pull request by MFreidank**
## Description

Adding an example for integration of custom handler as described in my blogpost:
https://medium.com/analytics-vidhya/deploy-huggingface-s-bert-to-production-with-pytorch-serve-27b068026d18

with the run_glue example (I'm restricting to the CoLA dataset in this example) given by transformers:
https://github.com/huggingface/transformers/blob/master/examples/run_glue.py

This example runs through in <30 minutes on a single Tesla K80 GPU (this can be further tweaked by training for less epochs, if needed).

Fixes #(issue)
Addresses issues #267 and #249

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Feature/Issue validation/testing

Please describe the tests [UT/IT] that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced. 
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- UT/IT execution results

- Logs

## Checklist:

- [ ] Have you added tests that prove your fix is effective or that this feature works?
- [x] New and existing unit tests pass locally with these changes?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

### CODE REVIEW

**Comment by MFreidank** [comment]
@harshbafna @mycpuorg 
As per our discussion on issue #267 and as follow-up to the earlier blogpost:
https://medium.com/analytics-vidhya/deploy-huggingface-s-bert-to-production-with-pytorch-serve-27b068026d18

**Comment by mycpuorg** [comment]
Thanks @MFreidank Putting this up for internal review and test.

**Comment by MFreidank** [comment]
Thank you @mycpuorg, let me know in case any changes or input is required.

File: examples/transformers_custom_handler/transformers_cola_handler.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            inputs = {input_name: input_values.to(self.device) for input_name, input_values in inputs.items()}
+            outputs, *_ = self.model(**inputs)
+
+        outputs = softmax(outputs.numpy())
+        prediction, = np.argmax(outputs, axis=1)
+
+        if self.mapping:
+            prediction = self.mapping[str(prediction)]
+
+        return [float(prediction)]
```
Comment (chauhang) [inline examples/transformers_custom_handler/transformers_cola_handler.py:85]: @MFreidank The return value is expected to be class string from the index_to_name.json mapping file. Please update logic for this

File: examples/transformers_custom_handler/download_glue_data.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+import os
+import sys
+import shutil
+import argparse
+import tempfile
+import urllib.request
+import zipfile
+
+TASKS = ["CoLA", "SST", "MRPC", "QQP", "STS", "MNLI", "SNLI", "QNLI", "RTE", "WNLI", "diagnostic"]
+TASK2PATH = {"CoLA":'https://firebasestorage.googleapis.com/v0/b/mtl-sentence-representations.appspot.com/o/data%2FCoLA.zip?alt=media&token=46d5e637-3411-4188-bc44-5809b5bfb5f4',
```
Comment (chauhang) [inline examples/transformers_custom_handler/download_glue_data.py:29]: @MFreidank Instead of hardcoding all the taskpath downloads, it will be better to use the pretrained models in the example. That way the example can be generic and one will not have to keep updating if the download path changes in future. Please see what can be done for generalizing the example

File: examples/transformers_custom_handler/README.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+The code for finetuning is based on the transformers GLUE CoLA example:
+https://github.com/huggingface/transformers/tree/v2.8.0/examples#glue
+
+Code for a simple custom handler for the resulting BERT model can be found in the 
+file `transformers_cola_handler.py`.
+
+
+# Finetuning
+
```
Comment (chauhang) [inline examples/transformers_custom_handler/README.md:14]: @MFreidank Please add all the required dependencies to the readme prerequisite install steps -- transformers and any others

File: examples/transformers_custom_handler/README.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    torch-model-archiver --model-name bert --version 1.0 --serialized-file outputs/pytorch_model.bin --handler "./transformers_cola_handler.py" --extra-files "./outputs/config.json,./outputs/vocab.txt"
+    ```
+
+ * Register the model on TorchServe using the above model archive file and run inference on a sample text
+   
+    ```bash
+    mkdir model_store
+    mv bert.mar model_store/
+    torchserve --start --model-store model_store --models bert=bert.mar
+    curl -X POST http://127.0.0.1:8080/predictions/bert -T sample_data/positive_cola_example
```
Comment (chauhang) [inline examples/transformers_custom_handler/README.md:37]: @MFreidank The Sample data file is missing in the PR. Please add that as well

File: examples/transformers_custom_handler/transformers_cola_handler.py (hunk 1 above)
Comment (MFreidank) [reply examples/transformers_custom_handler/transformers_cola_handler.py]: @chauhang Resolved. My example now makes use of "index_to_name.json", outputting verbose class names as prediction rather than just labels.

File: examples/transformers_custom_handler/transformers_cola_handler.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    def __init__(self):
+        super(TransformersClassifierHandler, self).__init__()
+        self.initialized = False
+
+    def initialize(self, ctx):
+        self.manifest = ctx.manifest
+
+        properties = ctx.system_properties
+        model_dir = properties.get("model_dir")
+        # self.device = torch.device("cuda:" + str(properties.get("gpu_id")) if torch.cuda.is_available() else "cpu")
```
[the remark this answers is not in the document]
Comment (MFreidank) [reply examples/transformers_custom_handler/transformers_cola_handler.py]: Resolved and tested that GPU serving works as intended on a Tesla K80 GPU.

File: examples/transformers_custom_handler/README.md (hunk 3 above)
Comment (MFreidank) [reply examples/transformers_custom_handler/README.md:17]: Resolved, added an `Install Dependencies` to my README that installs recommended versions of the libraries.

File: examples/transformers_custom_handler/README.md (hunk 4 above)
Comment (MFreidank) [reply examples/transformers_custom_handler/README.md]: Resolved, but changed data for easier usage of pretrained models. 
New sample data is: `sample_data/positive_sentiment_example.txt` and `sample_data/negative_sentiment_example.txt` which are two data snippets that come from the binary sentiment analysis dataset [SST-2](https://nlp.stanford.edu/sentiment/treebank.html).

File: examples/transformers_custom_handler/download_glue_data.py (hunk 2 above)
Comment (MFreidank) [reply examples/transformers_custom_handler/download_glue_data.py]: @chauhang Thanks for your feedback - very valid point. I've significantly simplified my example, I am loading a pretrained checkpoint using huggingface for a distilBERT model pre-trained on the SST-2 binary sentiment analysis task. 
This alleviates any need to hardcode paths or fine-tune any model, making the example significantly more accessible.

Model snapshot I am using can be found [here](https://huggingface.co/distilbert-base-uncased-finetuned-sst-2-english).

More information on the SST-2 task the model was finetuned for can be found [here](https://nlp.stanford.edu/sentiment/).

**Comment by MFreidank** [comment]
> @MFreidank Thanks for submitting the PR and a very nice blog article. I have left few comments in the code. Please take a look.

@chauhang Many thanks for your review - I was able to simplify the example quite a bit. 
I am now using a pre-trained transformers checkpoint of a distilBERT model for a binary sentiment classification task (SST2, Stanford Sentiment Treebank). 

The example can be tested in under 2 minutes now and runs on commodity hardware, with or without available GPU.

Could you check to see if there are any additional changes required?

**Comment by chauhang** [comment]
Thanks @MFreidank, please work with @HamidShojanazeri  to merge this PR with https://github.com/pytorch/serve/pull/302 for a more general and complete sol for handling BERT models.

**Comment by HamidShojanazeri** [comment]
Thanks @MFreidank for your great post, it was very helpful. I would appreciate if we could work together and integrate this PR with [#302](https://github.com/pytorch/serve/pull/302). I will send you an email, to discuss it over there.

**Comment by maaquib** [comment]
@MFreidank @HamidShojanazeri Thanks for this contribution. Given #302 is merged, is there any further work required on this PR or can we close it?

**Comment by maaquib** [comment]
After talking to @HamidShojanazeri it looks like these changes have been incorporated in #302 
Closing this. @MFreidank @HamidShojanazeri Please reopen if there is any additional changes here
